src/ec/Jacobi.mo

```diff
@@ -0,0 +1,375 @@
+    switch point {
+      case (#infinity) {
+        return #infinity;
+      };
+      case (#point (x, y, z, curve)) {
+        if (z.isEqual(curve.Fp(1))) {
+          return point;
+        };
+
+        let zInverse = z.inverse();
+        let zzInverse = zInverse.sqr();
+
+        let newX = x.mul(zzInverse);
+        let newY = y.mul(zzInverse).mul(zInverse);
+        let newZ = curve.Fp(1);
+
+        return #point (newX, newY, newZ, curve);
+      };
+    };
+  };
+
+  public func double(point : Point) : Point {
+    switch point {
+      case (#infinity) {
+        return #infinity;
+      };
+      case (#point (x, y, z, curve)) {
+        let (x2, y2, z2) = doDouble(x.value, y.value, z.value, curve.a, curve.p);
+        return #point (fpFromInt(x2, curve), fpFromInt(y2, curve),
+            fpFromInt(z2, curve), curve);
+      };
+    };
+  };
+
+  public func mul(point: Point, other: Nat) : Point {
+    if (other == 0) {
+      return #infinity;
+    };
+    switch (scale(point)) {
+      case (#infinity) {
+        return #infinity;
+      };
+      case (#point (x2, y2, _, curve)) {
+        var p : (Int, Int, Int) = (0, 0, 1);
+        let naf = Numbers.toNaf(other);
+
+        for (i in Iter.revRange(naf.size() - 1, 0)) {
+          let nafItem : Int = naf[Int.abs(i)];
+          p := doDouble(p.0, p.1, p.2, curve.a, curve.p);
+          if (nafItem < 0) {
+            p := _add(p.0, p.1, p.2, x2.value, y2.neg().value, 1,
+              curve.a, curve.p);
+          } else if (nafItem > 0) {
+            p := _add(p.0, p.1, p.2, x2.value, y2.value, 1, curve.a, curve.p);
+          };
+        };
+        if (p.1 == 0 or p.2 == 0) {
+          return #infinity;
+        };
+        return #point (fpFromInt(p.0, curve), fpFromInt(p.1, curve), fpFromInt(p.2, curve), curve);
+      };
+    };
+  };
+
+  public func mulBase(other : Nat, curve : Curves.Curve) : Point {
+    return mul(base(curve), other);
+  };
+
+  public func add(point1 : Point, point2: Point) : Point {
+    return switch point1 {
+      case (#infinity) {
+        point2
+      };
+      case (#point (X1, Y1, Z1, curve)) {
+        switch point2 {
+          case (#infinity) {
+            point1
+          };
+          case (#point (X2, Y2, Z2, _)) {
+            let (X3, Y3, Z3) = _add(X1.value, Y1.value, Z1.value,
+              X2.value, Y2.value, Z2.value, curve.a, curve.p);
+            if (Y3 == 0 or Z3 == 0) {
+              #infinity;
+            } else {
+              #point (fpFromInt(X3, curve), fpFromInt(Y3, curve),
+                fpFromInt(Z3, curve), curve);
+            };
+          };
+        };
+      };
+    };
+  };
+
+  func doDouble(X1 : Int, Y1 : Int, Z1 : Int, a : Nat, p : Nat) : (Int, Int, Int) {
+   if (Z1 == 1) {
+     return doubleWithZ1(X1, Y1, a, p)
+   };
+   if (Y1 == 0 or Z1 == 0) {
+     return (0, 0, 1);
+   };
+
+   let (XX, YY) = (X1 * X1 % p, Y1 * Y1 % p);
+   if (YY == 0) {
+     return (0, 0, 1);
```

Comment:
Yes I suppose it's possible. It's just an optimization where `Y1 == 0` is a cheap check to return early. The decisive check should be the more expensive `Y1 % p == 0`, but that gets inferred from  `YY` where the modular reduction happens any way.

Reply:
I think it's possible since in `Y1 == 0`, `Y1` is not reduced `% p`. In this case it's just a cheap comparison in order to return early and we'd still need to check for `YY == 0`.